packages/asset/contracts/AssetCreate.sol

```diff
@@ -0,0 +1,336 @@
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/EIP712Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "./libraries/TokenIdUtils.sol";
+import "./AuthValidator.sol";
+import "./ERC2771Handler.sol";
+import "./interfaces/IAsset.sol";
+import "./interfaces/ICatalyst.sol";
+import "./interfaces/IAssetCreate.sol";
+
+/// @title AssetCreate
+/// @author The Sandbox
+/// @notice User-facing contract for creating new assets
+contract AssetCreate is
+    IAssetCreate,
+    Initializable,
+    ERC2771Handler,
+    EIP712Upgradeable,
+    AccessControlUpgradeable
+{
+    using TokenIdUtils for uint256;
+
+    IAsset private assetContract;
+    ICatalyst private catalystContract;
+    AuthValidator private authValidator;
+
+    // mapping of creator address to creator nonce, a nonce is incremented every time a creator mints a new token
+    mapping(address => uint16) public creatorNonces;
+
+    bytes32 public constant SPECIAL_MINTER_ROLE =
+        keccak256("SPECIAL_MINTER_ROLE");
+    bytes32 public constant BRIDGE_MINTER_ROLE =
+        keccak256("BRIDGE_MINTER_ROLE");
+    bytes32 public constant MINT_TYPEHASH =
+        keccak256(
+            "Mint(address creator,uint16 nonce,uint8 tier,uint256 amount,bool revealed,string metadataHash)"
+        );
+    bytes32 public constant MINT_BATCH_TYPEHASH =
+        keccak256(
+            "Mint(address creator,uint16 nonce,uint8[] tiers,uint256[] amounts,bool[] revealed,string[] metadataHashes)"
+        );
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice Initialize the contract
+    /// @param _assetContract The address of the asset contract
+    /// @param _authValidator The address of the AuthValidator contract
+    /// @param _forwarder The address of the forwarder contract
+    function initialize(
+        string memory _name,
+        string memory _version,
+        address _assetContract,
+        address _catalystContract,
+        address _authValidator,
+        address _forwarder,
+        address _defaultAdmin
+    ) public initializer {
+        assetContract = IAsset(_assetContract);
+        catalystContract = ICatalyst(_catalystContract);
+        authValidator = AuthValidator(_authValidator);
+        __ERC2771Handler_initialize(_forwarder);
+        __EIP712_init(_name, _version);
+        __AccessControl_init();
+        _grantRole(DEFAULT_ADMIN_ROLE, _defaultAdmin);
+    }
+
+    /// @notice Create a new asset
+    /// @param signature A signature generated by TSB
+    /// @param tier The tier of the asset to mint
+    /// @param amount The amount of the asset to mint
+    /// @param metadataHash The metadata hash of the asset to mint
+    function createAsset(
+        bytes memory signature,
+        uint8 tier,
+        uint256 amount,
+        bool revealed,
+        string calldata metadataHash
```

Comment:
Need to add a check which limits metadataHash to the original (first) creator that used it

Reply:
It is already limited by the fact that every metadata hash can only be used once.

To retrieve the original creator we can use:
```
// mapping of ipfs metadata token hash to token ids
mapping(string => uint256) public hashUsed;`
```
and from the above tokenId you can get the creator.

Reply:
Ok, I thought we were allowing Creators to use the same hash more than once (but with a different tokenId) - maybe I missed the update

Reply:
I'll tag @rishabh0x00 as he was working on the metadata.

Reply:
Yeah, I remember that, and I think with the current implementation even the creator won't be able to mint twice, unless something will be different in the metadata, so for example the tier of the asset.

I wonder if thats a problem or not because the minting multiple times by the same creator was AFAIK a side-effect.

Reply:
Reminting by same creator with same tokenId - I have tagged you in the googledoc thread on this topic. I understood that the requirements are: a creator can reuse the same metadata hash (ie no changes needed to metadata at all) for minting under a new tokenId; a different creator cannot use this hash.

Let me know if this is out of date by sharing the relevant thread. Thanks!

Reply:
@rishabh0x00 see above. Will you be making a separate PR?

Reply:
From requirements doc at https://docs.google.com/document/d/1DINSg9Kac7GPZIz8gtGv2u3VHPLwo36Ume4omnUDKes/ :
--> creator can reuse a metadata hash
--> a different creator can't use someone else's hash that's already been minted
--> if a creator wants to use the same hash more than once - even with the same tier - it's allowed (but just stupid commercially) - with a new tokenId